### PR TITLE
charts - generalize evaluating expressions for charts

### PR DESCRIPTION
**Pull request by crnjan**
This is based on https://github.com/openhab/openhab-webui/pull/1419. Also contains fixes from https://github.com/openhab/openhab-webui/pull/1634.

fixes https://github.com/openhab/openhab-webui/issues/1418
fixes https://github.com/openhab/openhab-webui/issues/1295
fixes https://github.com/openhab/openhab-webui/issues/1102
fixes https://github.com/openhab/openhab-webui/issues/1072

It's still WIP, need to perform more tests, any help greatly appreciated 

EDIT: went through cases reported in above issues and seems to work now. Also, as far as I tested (so far) locally with my setup seems to work - but probably don't have all use-cases setup so probably missed something - therefore would be much appreciated if any volunteer tries it out too, thank you!

### CODE REVIEW

**Comment by ghys** [comment]
Thanks! Generalizing expressions in chart widgets is a plus for sure! It's looking good code wise but I'll have a test run soon.

**Comment by crnjan** [comment]
Created a test jar, please run:
```
bundle:update org.openhab.ui https://github.com/crnjan/openhab-webui/raw/feature/diff_poc/bundles/org.openhab.ui/target/org.openhab.ui-4.0.0-SNAPSHOT_eval.jar
```
in order to try it out. If there is any case not working, please let me know - and provide a (minimal) working sample (yaml) so I can try locally.

If needed, revert back to official UI jar with

```
bundle:update org.openhab.ui https://ci.openhab.org/job/openHAB-WebUI/lastSuccessfulBuild/artifact/bundles/org.openhab.ui/target/org.openhab.ui-4.0.0-SNAPSHOT.jar
```

**Comment by crnjan** [comment]
Accidentally pushed (potential) fix for https://github.com/openhab/openhab-webui/issues/1072, introducing `onStates`

```
        markArea:
          - component: oh-mark-area
            config:
              item: gBasement_BoilerRoom_HP_Compressor_Powered
              onStates: OFF 
```

I can revert if needed ...

**Comment by ghys** [comment]
> I can revert if needed ... also feel free to suggest a better name ...

It's sort-of related in a general sense (allow customizing charts beyond some arbitrary/common sense defaults, for those who need this customization), so no need to open a separate PR. The name `onStates` is fine, maybe `markStates` alternatively or simply `on` or `states`. Not sure which I like better? Since oh-mark-area isn't currently described and won't appear in config sheets or https://next.openhab.org/docs/ui/components/#chart-page-components (probably something to remedy one day) it's kind of obscure anyway :)

By the way it accidentally works when you feed that parameter a string instead of an array (because `includes` is both in the prototype of Array and String) but it might not work 100% as the user expects (for instance `onStates: ONOFF` will match both `ON` and `OFF`, also `onStates: ="ABC123"` will match `AB`, `BC`, `C12`... but `onStates: 1` won't match `1`, if needs to be `onStates: [1]`). Maybe transforming it into a single-element array if it's not an array would be safest.

NB: I tried this with a scene Number item but noticed it's actually a string in the API response in this format in my case:

image

Since `"4.0" == 4` in JS maybe it would be valuable to use the non-strict equality somehow to check if there's a match?

File: bundles/org.openhab.ui/web/src/components/widgets/chart/series/oh-mark-area.js [hunk 1]
```diff
@@ -1,20 +1,22 @@
...
 import LocalizedFormat from 'dayjs/plugin/localizedFormat'
+import ComponentId from '../../component-id'
 dayjs.extend(LocalizedFormat)
 
 export default {
   get (component, points, startTime, endTime, chart) {
-    const itemPoints = points.find(p => p.name === component.config.item).data
-    let markArea = Object.assign({}, component.config)
+    let markArea = chart.evaluateExpression(ComponentId.get(component), component.config)
+    const onStates = markArea.onStates || ['ON', 'OPEN']
```
Comment (ghys) [inline bundles/org.openhab.ui/web/src/components/widgets/chart/series/oh-mark-area.js:9]:
```suggestion
    const onStates = (markArea.onStates && !Array.isArray(markArea.onStates) ? [markArea.onStates] : markArea.onStates) || ['ON', 'OPEN']
```

File: bundles/org.openhab.ui/web/src/components/widgets/chart/series/oh-mark-area.js [hunk 2]
```diff
@@ -1,20 +1,22 @@
...
     markArea.data = []
     let rollingState = null
     let currentArea = null
     itemPoints.forEach((p) => {
-      if ((p.state === 'ON' && rollingState !== 'ON') || (p.state === 'OPEN' && rollingState !== 'OPEN')) {
-        currentArea = [{ name: component.config.name, xAxis: new Date(p.time) }]
-      } else if (((p.state !== 'ON' && rollingState === 'ON') || (p.state !== 'OPEN' && rollingState === 'OPEN')) && currentArea) {
+      if (onStates.includes(p.state) && !onStates.includes(rollingState)) {
+        currentArea = [{ name: markArea.name, xAxis: new Date(p.time) }]
+      } else if ((!onStates.includes(p.state) && onStates.includes(rollingState)) && currentArea) {
```
Comment (ghys) [inline bundles/org.openhab.ui/web/src/components/widgets/chart/series/oh-mark-area.js:17]:
```suggestion
      // eslint-disable-next-line eqeqeq
      if (onStates.findIndex((s) => s == p.state) >= 0 && onStates.findIndex((s) => s == rollingState) < 0) {
        currentArea = [{ name: markArea.name, xAxis: new Date(p.time) }]
      // eslint-disable-next-line eqeqeq
      } else if ((onStates.findIndex((s) => s == p.state) < 0 && onStates.findIndex((s) => s == rollingState) >= 0) && currentArea) {
```

**Review by ghys — COMMENTED** [review]
Suggestions for the above comment.

The rest of the PR LGTM!

**Pull request closed without merging**